connector/linkedin/linkedin.go

```diff
@@ -0,0 +1,181 @@
+// Package linkedin provides authentication strategies using LinkedIn
+package linkedin
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"golang.org/x/oauth2"
+
+	"github.com/coreos/dex/connector"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	apiURL   = "https://api.linkedin.com/v1"
+	authURL  = "https://www.linkedin.com/oauth/v2/authorization"
+	tokenURL = "https://www.linkedin.com/oauth/v2/accessToken"
+)
+
+// Config holds configuration options for LinkedIn logins.
+type Config struct {
+	ClientID     string `json:"clientID"`
+	ClientSecret string `json:"clientSecret"`
+	RedirectURI  string `json:"redirectURI"`
+}
+
+// Open returns a strategy for logging in through LinkedIn
+func (c *Config) Open(logger logrus.FieldLogger) (connector.Connector, error) {
+	return &linkedInConnector{
+		oauth2Config: &oauth2.Config{
+			ClientID:     c.ClientID,
+			ClientSecret: c.ClientSecret,
+			Endpoint: oauth2.Endpoint{
+				AuthURL:  authURL,
+				TokenURL: tokenURL,
+			},
+			Scopes:      []string{"r_basicprofile", "r_emailaddress"},
+			RedirectURL: c.RedirectURI,
+		},
+		logger: logger,
+	}, nil
+}
+
+type connectorData struct {
+	AccessToken string `json:"accessToken"`
+}
+
+type linkedInConnector struct {
+	oauth2Config *oauth2.Config
+	logger       logrus.FieldLogger
+}
+
+// LinkedIn doesn't provide refresh tokens, so refresh tokens issued by Dex
+// will expire in 60 days (default LinkedIn token lifetime).
+var (
+	_ connector.CallbackConnector = (*linkedInConnector)(nil)
+	_ connector.RefreshConnector  = (*linkedInConnector)(nil)
+)
+
+// LoginURL returns an access token request URL
+func (c *linkedInConnector) LoginURL(scopes connector.Scopes, callbackURL, state string) (string, error) {
+	if c.oauth2Config.RedirectURL != callbackURL {
+		return "", fmt.Errorf("expected callback URL %q did not match the URL in the config %q",
+			callbackURL, c.oauth2Config.RedirectURL)
+	}
+
+	return c.oauth2Config.AuthCodeURL(state), nil
+}
+
+// HandleCallback handles HTTP redirect from LinkedIn
+func (c *linkedInConnector) HandleCallback(s connector.Scopes, r *http.Request) (identity connector.Identity, err error) {
+	q := r.URL.Query()
+	if errType := q.Get("error"); errType != "" {
+		return identity, &oauth2Error{errType, q.Get("error_description")}
+	}
+
+	ctx := r.Context()
+	token, err := c.oauth2Config.Exchange(ctx, q.Get("code"))
+	if err != nil {
+		return identity, fmt.Errorf("linkedin: get token: %v", err)
+	}
+
+	client := c.oauth2Config.Client(ctx, token)
+	profile, err := c.profile(ctx, client)
+	if err != nil {
+		return identity, fmt.Errorf("github: get profile: %v", err)
+	}
+
+	identity = connector.Identity{
+		UserID:        profile.ID,
+		Username:      profile.FirstName + " " + profile.LastName,
+		Email:         profile.Email,
```

Comment:
Please error if profile.Email isn't set.

Reply:
That's seems like an impossible condition. This is the main email that was used for registration and we always ask permission for r_emailaddress claim.
I'd add the check just to be sure anyway.